torch/utils/tensorboard/writer.py

```diff
@@ -0,0 +1,731 @@
+        return self.file_writer
+
+    def add_scalar(self, tag, scalar_value, global_step=None, walltime=None):
+        """Add scalar data to summary.
+
+        Args:
+            tag (string): Data identifier
+            scalar_value (float or string/blobname): Value to save
+            global_step (int): Global step value to record
+            walltime (float): Optional override default walltime (time.time()) of event
+        """
+        if self._check_caffe2_blob(scalar_value):
+            scalar_value = workspace.FetchBlob(scalar_value)
+        self.get_file_writer().add_summary(
+            scalar(tag, scalar_value), global_step, walltime)
+
+    def add_scalars(self, main_tag, tag_scalar_dict, global_step=None, walltime=None):
+        """Adds many scalar data to summary.
+
+        Note that this function also keeps logged scalars in memory. In extreme case it explodes your RAM.
+
+        Args:
+            main_tag (string): The parent name for the tags
+            tag_scalar_dict (dict): Key-value pair storing the tag and corresponding values
+            global_step (int): Global step value to record
+            walltime (float): Optional override default walltime (time.time()) of event
+
+        Examples::
+
+            writer.add_scalars('run_14h', {'xsinx':i*np.sin(i/r),
+                                           'xcosx':i*np.cos(i/r),
+                                           'arctanx': numsteps*np.arctan(i/r)}, i)
+            # This call adds three values to the same scalar plot with the tag
+            # 'run_14h' in TensorBoard's scalar section.
+        """
+        walltime = time.time() if walltime is None else walltime
+        fw_logdir = self.get_file_writer().get_logdir()
+        for tag, scalar_value in tag_scalar_dict.items():
+            fw_tag = fw_logdir + "/" + main_tag + "/" + tag
+            if fw_tag in self.all_writers.keys():
+                fw = self.all_writers[fw_tag]
+            else:
+                fw = FileWriter(logdir=fw_tag)
+                self.all_writers[fw_tag] = fw
+            if self._check_caffe2_blob(scalar_value):
+                scalar_value = workspace.FetchBlob(scalar_value)
+            fw.add_summary(scalar(main_tag, scalar_value),
+                           global_step, walltime)
+            self.__append_to_scalar_dict(
+                fw_tag, scalar_value, global_step, walltime)
+
+    def export_scalars_to_json(self, path):
+        """Exports to the given path an ASCII file containing all the scalars written
+        so far by this instance, with the following format:
+        {writer_id : [[timestamp, step, value], ...], ...}
+
+        The scalars saved by ``add_scalars()`` will be flushed after export.
+        """
+        with open(path, "w") as f:
+            json.dump(self.scalar_dict, f)
+        self.scalar_dict = {}
+
+    def add_histogram(self, tag, values, global_step=None, bins='tensorflow', walltime=None, max_bins=None):
```

Comment:
Is 'tensorflow' really an appropriate default here?

Reply:
I didn't have time to figure it out, but the default should match whatever `torch.histc` does.

Reply:
Actually, is the real problem that you want `add_histogram` to automatically compute bins, and torch.histc doesn't do that, and therefore you're going to go about doing this some other way? I guess for a visualization application, automatic bin computation is reasonable. But I'd also at least offer the PyTorch style, where you can explicitly specify min/max/bins explicitly

Reply:
The tensorflow default is a set of predefined bins. https://github.com/lanpa/tensorboardX/blob/0bf6c07d97b0745654fd9fab8ee3261ec707f253/tensorboardX/writer.py#L221 The distance between two neighboring bins is proportional to their distance to zero. In other words, the resolution near zero is very good in such setting. I guess it also helps the debugging since most weight value doesn't drift to big numbers (If your network doesn't go wild). Another benefit is the speed, since the bins are only computed once during initialization, you only need to scan the tensor values once for each add_histogram call. Finally, computing bins dynamically might be harder to debug; In the early stage of tensorboardX implementation, I use `np.histogram`'s binning method as default and got many issues feedback about histogram plot. Then I followed the tensorflow implementation and there is no complaint. For visualization, the unused bins will be removed before the protobuf is written to the disk, so there is no need to worry about that histogram plot been occupied empty bins near +inf and -inf. min/max/bins can be specified manually just like the way np.histogram did. For example: `bins=numpy.arange(1, 2, 0.1)`